File: scala/src/test/scala/scalasearch/FileUtilTest.scala
package scalasearch

import org.junit.runner.RunWith
import org.scalatest.FunSuite
import org.scalatest.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
class FileUtilTest extends FunSuite {
  /***************************************************************************
    * getExtension tests
    **************************************************************************/
  test("test get txt extension") {
    assert(FileUtil.getExtension(".filename.txt").equalsIgnoreCase("txt"))
  }

  test("test get TXT extension") {
    assert(FileUtil.getExtension(".filename.TXT").equalsIgnoreCase("txt"))
  }

  test("test get missing extension") {
    assert(FileUtil.getExtension("filename.") == "")
  }

  test("test get no extension") {
    assert(FileUtil.getExtension("filename") == "")
  }

  test("test hidden get txt extension") {
    assert(FileUtil.getExtension(".filename.txt").equalsIgnoreCase("txt"))
  }

  test("test hidden get missing extension") {
    assert(FileUtil.getExtension(".filename.") == "")
  }

  test("test hidden get no extension") {
    assert(FileUtil.getExtension(".filename") == "")
  }

  /***************************************************************************
    * isDotDir tests
    **************************************************************************/
  test("test isDotDir single dot") {
    assert(FileUtil.isDotDir("."))
  }

  test("test isDotDir double dot") {
    assert(FileUtil.isDotDir(".."))
  }

  test("test isDotDir hidden file") {
    assert(!FileUtil.isDotDir(".gitignore"))
  }

  /***************************************************************************
    * isHidden tests
    **************************************************************************/
  test("test isHidden single dot") {
    assert(!FileUtil.isHidden("."))
  }

  test("test isHidden double dot") {
    assert(!FileUtil.isHidden(".."))
  }

  test("test isHidden hidden file name") {
    assert(FileUtil.isHidden(".gitignore"))
  }

  test("test isHidden non-hidden file name") {
    assert(!FileUtil.isHidden("filename.txt"))
  }

  test("test isHidden non-hidden file path with dot") {
    assert(!FileUtil.isHidden("./filename.txt"))
  }

  /***************************************************************************
    * splitPath tests
    **************************************************************************/
  test("test splitPath path with dot") {
    val path = "./path/to/somewhere/"
    val elems = FileUtil.splitPath(path)
    assert(elems.size == 3)
    assert(elems.head == "path")
  }

  test("test splitPath path with double dot") {
    val path = "../path/to/somewhere/"
    val elems = FileUtil.splitPath(path)
    assert(elems.size == 3)
    assert(elems.head == "path")
  }

  test("test splitPath path without dot") {
    val path = "/path/to/somewhere/"
    val elems = FileUtil.splitPath(path)
    println("elems:" + elems)
    assert(elems.size == 3)
    assert(elems.head == "path")
  }
}

File: dart/dartsearch/test/file_util_test.dart
import 'package:test/test.dart';

void main() {
  /***************************************************************************
   * isDotDir tests
   **************************************************************************/
  group('isDotDir tests', () {
    test('test isDotDir single dot', () {
      expect(FileUtil.isDotDir('.'), true);
    });

    test('test isDotDir double dot', () {
      expect(FileUtil.isDotDir('..'), true);
    });

    test('test isDotDir path without dot', () {
      expect(FileUtil.isDotDir('~/path'), false);
    });

    test('test isDotDir path with dot', () {
      expect(FileUtil.isDotDir('./path'), false);
    });

    test('test isDotDir hidden file', () {
      expect(FileUtil.isDotDir('.gitignore'), false);
    });
  });

  /***************************************************************************
   * isHidden tests
   **************************************************************************/
  group('isHidden tests', () {
    test('test isHidden single dot', () {
      expect(FileUtil.isHidden('.'), false);
    });

    test('test isHidden double dot', () {
      expect(FileUtil.isHidden('..'), false);
    });

    test('test isHidden hidden file', () {
      expect(FileUtil.isHidden('.gitignore'), true);
    });

    test('test isHidden not-hidden file', () {
      expect(FileUtil.isHidden('filename.txt'), false);
    });
  });

  /***************************************************************************
   * expandPath tests
   **************************************************************************/
  group('expandPath tests', () {
    var homePath = FileUtil.homePath();
    test('test expandPath path with tilde', () {
      expect(FileUtil.expandPath('~/path'), '$homePath/path');
    });

    test('test expandPath path tilde only', () {
      expect(FileUtil.expandPath('~'), homePath);
    });

    test('test expandPath relative path no tilde', () {
      expect(FileUtil.expandPath('./path'), './path');
    });

    test('test expandPath absolute home path no tilde', () {
      expect(FileUtil.expandPath('$homePath/path'), '$homePath/path');
    });
  });

  /***************************************************************************
   * extension tests
   **************************************************************************/
  group('extension tests', () {
    test('test extension file with extension', () {
      expect(FileUtil.extension('filename.txt'), 'txt');
    });

    test('test extension file no extension', () {
      expect(FileUtil.extension('README'), '');
    });

    test('test extension file empty extension', () {
      expect(FileUtil.extension('filewithdot.'), '');
    });

    test('test extension hidden file with extension', () {
      expect(FileUtil.extension('.hidden.txt'), 'txt');
    });

    test('test extension hidden file no extension', () {
      expect(FileUtil.extension('.gitignore'), '');
    });
  });
}
